Mathlib/Order/Quantale.lean

```diff
@@ -0,0 +1,159 @@
+semigroup. As we define quantales relative to their semigroup, these do not need to be defined
+explicitly here. An integral (or strictly two-sided) quantale is a unital quantale in which
+the top element of the lattice and the unit of the semigroup coincide. We give a mix-in class
+definition for this.
+
+The involutive quantale (which is necessary to discuss regularity properties) we do not cover
+in this file. Also the proof that every frame is a commutative quantale, and that a quantale is
+a frame if and only if it is integral and idempotent are developed separately in order to reduce
+overhead if a user does not need them.
+
+## Main definitions
+
+* class `Quantale`: a semigroup distributing over a complete lattice, i.e satisfying
+  `x * (sSup s) = ⨆ y ∈ s, x * y` and `(sSup s) * y = ⨆ x ∈ s, x * y`;
+
+* mix-in class `IsIntegral`: a unital quantale (i.e. a quantale in which the semigroup is
+  a monoid) is called integral, or strictly two-sided, when `⊤ = 1`;
+
+* next to these classes, we define the additive versions `AddQuantale`, `IsAddIntegral` in
+  which the semigroup is denoted `+` i.s.o. `*`;
+
+## Naming conventions
+
+## Notation
+
+* `x ⇨ₗ y` : `sSup { z | z * x ≤ y }`, the left-residuation of `y` over `x`;
+
+* `x ⇨ᵣ y` : `sSup { z | x * z ≤ y }`, the right-residuation of `y` over `x`;
+
+## References
+
+<https://en.wikipedia.org/wiki/Quantale>
+<https://encyclopediaofmath.org/wiki/Quantale>
+<https://ncatlab.org/nlab/show/quantale>
+
+## TODO
+
+-/
+
+/-- An additive quantale is an additive semigroup distributing over a complete lattice. -/
+class AddQuantale (α : Type*) [AddSemigroup α] extends CompleteLattice α where
+  /-- Addition is distributive over join in a quantale -/
+  protected add_sSup_eq_iSup_add (x : α) (s : Set α) : x + sSup s = ⨆ y ∈ s, x + y
+  /-- Addition is distributive over join in a quantale -/
+  protected sSup_add_eq_iSup_add (s : Set α) (y : α) : sSup s + y = ⨆ x ∈ s, x + y
+
+/-- A quantale is a semigroup distributing over a complete lattice. -/
+@[to_additive]
+class Quantale (α : Type*) [Semigroup α] extends CompleteLattice α where
+  /-- Multiplication is distributive over join in a quantale -/
+  protected mul_sSup_eq_iSup_mul (x : α) (s : Set α) : x * sSup s = ⨆ y ∈ s, x * y
+  /-- Multiplication is distributive over join in a quantale -/
+  protected sSup_mul_eq_iSup_mul (s : Set α) (y : α) : sSup s * y = ⨆ x ∈ s, x * y
+
+/-- An integral (or strictly two-sided) additive quantale is a quantale over an additive monoid
+    where top and unit coincide. -/
+class IsAddIntegral (α : Type*) [AddMonoid α] [AddQuantale α] : Prop where
+  /-- Top and unit coincide in an integral (or strictly two-sided) quantale -/
+  protected top_eq_zero : (⊤ : α) = 0
+
+/-- An integral (or strictly two-sided) quantale is a quantale over a monoid where
+    top and unit coincide. -/
+@[to_additive]
+class IsIntegral (α : Type*) [Monoid α] [Quantale α] : Prop where
```

Comment:
This name is way too generic to be placed in the root namespace. Can you rename to something like `IsIntegralQuantale` or `Quantal.IsIntegral`? (slight preference for the first)

Feel free to complain on Zulip that the ring theoretic `IsIntegral` is not namespace, though 😉

Reply:
Renaming it to Quantale.IsIntegral is not a problem.

Renaming it to IsIntegralQuantale leads to the additive version being IsAddIntegralAddQuantale, which I think is a bit ugly. That can probably be solved in the Frontend of ToAdditive, but I don't know how exactly.

Why do you have a slight preference for IsIntegralQuantale?
(And how about the IsIdem for Semigroups? I guess that should follow the same approach then?)

Reply:
PS. I can for the time being leave out the Kleene changes, which makes for a smaller PR, but I'm struggling with GIT to do the revert in a nice way. Should I do that, or is it okay to make the update to Kleene in the same PR?

Reply:
> Renaming it to IsIntegralQuantale leads to the additive version being IsAddIntegralAddQuantale

Shouldn't the additive version be `IsIntegralAddQuantale`? Note that you can tweak the name that `to_additive` guesses

Reply:
Got it right now. I updated to_additive earlier, undid that change now, so now your suggestion makes more sense. I've updated it and also applied it to the IsIdemSemigroup.